### PR TITLE
autonat: add metrics

### PR DESCRIPTION
**Pull request by sukunrt**
fixes: https://github.com/libp2p/go-libp2p/issues/2017

### CODE REVIEW

**Comment by sukunrt** [comment]
@marten-seemann I've added. 
```
Current Reachability Status
Current Reachability Confidence
Client: DialResponses by type
Server: DialResponses by type
Server: DialRejections by type(currently the only case handled is rate limited)
```

File: p2p/host/autonat/grafana-dashboards/autonat.json [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+                                        "editorMode": "code",
+                                        "exemplar": false,
+                                        "expr": "increase(libp2p_autonat_client_dialresponse_total[$__rate_interval])",
+                                        "instant": false,
+                                        "legendFormat": "{{response_status}}",
+                                        "range": true,
+                                        "refId": "A"
+                                }
+                        ],
+                        "title": "Client: DialResponse by Type",
```
Comment (marten-seemann) [inline p2p/host/autonat/grafana-dashboards/autonat.json:362]:
```suggestion
                        "title": "Received Dial Responses",
```

File: p2p/host/autonat/grafana-dashboards/autonat.json [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+                                                "uid": "${DS_PROMETHEUS}"
+                                        },
+                                        "editorMode": "code",
+                                        "expr": "increase(libp2p_autonat_server_dialrefused_total[$__rate_interval])",
+                                        "legendFormat": "{{refusal_reason}}",
+                                        "range": true,
+                                        "refId": "A"
+                                }
+                        ],
+                        "title": "Server: Dial requests rate limited",
```
Comment (marten-seemann) [inline p2p/host/autonat/grafana-dashboards/autonat.json:470]: This doesn't seem to work?

image

File: p2p/host/autonat/grafana-dashboards/autonat.json [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+                                                "uid": "${DS_PROMETHEUS}"
+                                        },
+                                        "editorMode": "code",
+                                        "expr": "increase(libp2p_autonat_server_dialresponse_total[$__rate_interval])",
+                                        "legendFormat": "{{response_status}}",
+                                        "range": true,
+                                        "refId": "A"
+                                }
+                        ],
+                        "title": "server: DialResponse By Type",
```
Comment (marten-seemann) [inline p2p/host/autonat/grafana-dashboards/autonat.json:638]: I have to admit, I'm always confused who's the client and server when thinking about AutoNAT. Maybe "Outgoing Dials" would be harder to mix up?

**Review by marten-seemann — COMMENTED** [review]
https://github.com/ipfs/kubo/pull/9647 was merged, so I could run this on my Kubo node. I had to rebase this branch on top of the current go-libp2p master first though.

This looks mostly good to me. I'm wondering if we should export the timestamp of the next probe (as calculated by `scheduleProbe` as well), might be nice to display that in Grafana as well.

File: p2p/host/autonat/grafana-dashboards/autonat.json (hunk 3 above)
Comment (sukunrt) [reply p2p/host/autonat/grafana-dashboards/autonat.json:638]: outgoing does seem better. will fix.

File: p2p/host/autonat/grafana-dashboards/autonat.json [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+                                                                        "index": 0,
+                                                                        "text": "Unknown"
+                                                                },
+                                                                "1": {
+                                                                        "color": "green",
+                                                                        "index": 1,
+                                                                        "text": "Public"
+                                                                },
+                                                                "2": {
+                                                                        "color": "red",
```
[the remark this answers is not in the document]
Comment (sukunrt) [reply p2p/host/autonat/grafana-dashboards/autonat.json:104]: that's fair. violet seems good to me.

File: p2p/host/autonat/grafana-dashboards/autonat.json (hunk 2 above)
Comment (marten-seemann) [reply p2p/host/autonat/grafana-dashboards/autonat.json:470]: Just checked my dashboard, and now it's showing one event (after a couple of hours), so it seems to work.

Why is this not part of the dashboard to the right though?

Comment (sukunrt) [reply p2p/host/autonat/grafana-dashboards/autonat.json:470]: The dashboard on the right has segregation by Message_ResponseStatus. So this point would be counted under dial refused. It does seem fine to separate rate limit and dial refused and display on the same dashboard. Should I do that?

File: p2p/host/autonat/grafana-dashboards/autonat.json (hunk 3 above)
Comment (sukunrt) [reply p2p/host/autonat/grafana-dashboards/autonat.json:638]: this is fixed.

File: p2p/host/autonat/grafana-dashboards/autonat.json (hunk 2 above)
Comment (sukunrt) [reply p2p/host/autonat/grafana-dashboards/autonat.json:470]: Added other Refusal reasons too.

File: p2p/host/autonat/grafana-dashboards/autonat.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (sukunrt) [reply p2p/host/autonat/grafana-dashboards/autonat.json:104]: fixed

File: p2p/host/autonat/grafana-dashboards/autonat.json [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+                                                "uid": "${DS_PROMETHEUS}"
+                                        },
+                                        "editorMode": "builder",
+                                        "expr": "libp2p_autonat_reachability_status_confidnce",
+                                        "legendFormat": "__auto",
+                                        "range": true,
+                                        "refId": "A"
+                                }
+                        ],
+                        "title": "Reachability status confidence",
```
[the remark this answers is not in the document]
Comment (sukunrt) [reply p2p/host/autonat/grafana-dashboards/autonat.json:231]: fixed

**Review by marten-seemann — APPROVED** [review]
Thank you for the great work @sukunrt!

**Pull request merged**